### DISCUSSION TITLE
How can I access theia-cloud services deployed on the Ubuntu server through a browser in other desktop environments on the LAN

### DISCUSSION THREAD

**Discussion by inlann** [eclipse-theia/theia-cloud#232] [Q&A] [answered]
Hello:

I deployed the Theia-cloud in minikube on an Ubuntu server using docker as the driver:

```bash
tptuser@theia-cloud:~/work/theia-cloud/terraform/configurations/minikube_getting_started$ export TF_LOG=Debug
tptuser@theia-cloud:~/work/theia-cloud/terraform/configurations/minikube_getting_started$ terraform apply
...
2023-08-22T09:59:47.684Z [DEBUG] provider: plugin process exited: path=.terraform/providers/registry.terraform.io/mrparkers/keycloak/4.2.0/linux_amd64/terraform-provider-keycloak_v4.2.0 pid=1987072
2023-08-22T09:59:47.684Z [DEBUG] provider: plugin exited

Apply complete! Resources: 8 added, 0 changed, 1 destroyed.

Outputs:

try_now = "https://192.168.49.2.nip.io/trynow/"
```

The `theia-cloud` is the hostname of my Ubuntu server. The ip of the minikube is `192.168.49.2` and pods are runing:

```bash
tptuser@theia-cloud:~/work/theia-cloud/terraform/configurations/minikube_getting_started$ minikube ip
192.168.49.2
tptuser@theia-cloud:~/work/theia-cloud/terraform/configurations/minikube_getting_started$ kubectl get pods -n theiacloud 
NAME                                       READY   STATUS    RESTARTS   AGE
landing-page-deployment-7d6f787d45-x2kzh   1/1     Running   0          17h
operator-deployment-5f7689b64c-jnz7r       1/1     Running   0          17h
operator-deployment-5f7689b64c-qssq7       1/1     Running   0          17h
service-deployment-6b9f4d4b86-d596x        1/1     Running   0          17h
tptuser@theia-cloud:~/work/theia-cloud/terraform/configurations/minikube_getting_started$ kubectl get ingress -n theiacloud 
NAME                          CLASS   HOSTS                 ADDRESS        PORTS     AGE
landing-page-ingress          nginx   192.168.49.2.nip.io   192.168.49.2   80, 443   17h
service-ingress               nginx   192.168.49.2.nip.io   192.168.49.2   80, 443   17h
theia-cloud-demo-ws-ingress   nginx   192.168.49.2.nip.io   192.168.49.2   80, 443   17h
```

The DNS resolve the try_now to the minikube:

```bash
tptuser@theia-cloud:~/work/theia-cloud/terraform/configurations/minikube_getting_started$ dig 192.168.49.2.nip.io

; <<>> DiG 9.18.12-0ubuntu0.22.04.2-Ubuntu <<>> 192.168.49.2.nip.io
;; global options: +cmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: NOERROR, id: 21481
;; flags: qr rd ra; QUERY: 1, ANSWER: 1, AUTHORITY: 0, ADDITIONAL: 1

;; OPT PSEUDOSECTION:
; EDNS: version: 0, flags:; udp: 65494
;; QUESTION SECTION:
;192.168.49.2.nip.io.		IN	A

;; ANSWER SECTION:
192.168.49.2.nip.io.	3600	IN	A	192.168.49.2

;; Query time: 1800 msec
;; SERVER: 127.0.0.53#53(127.0.0.53) (UDP)
;; WHEN: Wed Aug 23 03:53:21 UTC 2023
;; MSG SIZE  rcvd: 64
```

It seems like the Theia-cloud works fine!

The IP of the Ubuntu host is `192.168.2.88`:

```bash
tptuser@theia-cloud:~/work/theia-cloud/terraform/configurations/minikube_getting_started$ ifconfig | grep inet
        inet 192.168.49.1  netmask 255.255.255.0  broadcast 192.168.49.255
        inet6 fe80::42:4dff:febc:5f68  prefixlen 64  scopeid 0x20<link>
        inet 172.17.0.1  netmask 255.255.0.0  broadcast 172.17.255.255
        inet6 fe80::42:2aff:fed4:e0f9  prefixlen 64  scopeid 0x20<link>
        inet 192.168.2.88  netmask 255.255.255.0  broadcast 192.168.2.255
        inet6 fe80::c482:d7ff:fee6:f54d  prefixlen 64  scopeid 0x20<link>
        inet 127.0.0.1  netmask 255.0.0.0
        inet6 ::1  prefixlen 128  scopeid 0x10<host>
        inet6 fe80::4023:a7ff:fe06:c3a  prefixlen 64  scopeid 0x20<link>
```

I'm trying to access the try_now url in the Chrome on my iMac since the Ubuntu server does not have any desktop environments. The IP of my iMac is `192.168.2.21`:

```bash
(base) mac@Titans-iMac ~ % ifconfig | grep inet
	inet 127.0.0.1 netmask 0xff000000 
	inet6 ::1 prefixlen 128 
	inet6 fe80::1%lo0 prefixlen 64 scopeid 0x1 
	inet6 fe80::1c35:89:22a2:f3b7%en0 prefixlen 64 secured scopeid 0x4 
	inet 192.168.2.21 netmask 0xffffff00 broadcast 192.168.2.255
	inet6 fe80::1033:93ff:cccd:1617%en1 prefixlen 64 secured scopeid 0x5 
	inet 192.168.0.81 netmask 0xffffff00 broadcast 192.168.0.255
	inet6 fe80::c2d:1ff:fe34:5a3c%awdl0 prefixlen 64 scopeid 0x6 
	inet6 fe80::c2d:1ff:fe34:5a3c%llw0 prefixlen 64 scopeid 0x9 
	inet6 fe80::378e:e6a2:55c:4eaf%utun0 prefixlen 64 scopeid 0xb 
	inet6 fe80::68c1:92f1:af40:9cc8%utun1 prefixlen 64 scopeid 0xc 
	inet6 fe80::ce81:b1c:bd2c:69e%utun2 prefixlen 64 scopeid 0xd 
	inet6 fe80::bac7:3f30:d7eb:201a%utun3 prefixlen 64 scopeid 0x10 
	inet6 fe80::892c:64eb:e2dc:e527%utun4 prefixlen 64 scopeid 0x11
```

I added `192.168.2.88 192.168.49.2.nip.io` to the `/etc/hosts` of the iMac but I cannot access the `https://192.168.2.88 192.168.49.2.nip.io` in the Chrome.

**Comment by inlann** [comment]
I tried to deploy a Nginx on the Ubuntu server. However, the Nginx configuration that brokers port 443 requires ssl authentication and a key. How can I get the certification and key from the Theia-cloud cluster? Or is there another way to access the try_now on another machine?

**Comment by jfaltermeier** [comment, marked as answer]
Hi, I think this is more of a minikube question than a Theia Cloud question. 
https://minikube.sigs.k8s.io/docs/start/ mentions that "minikube is local Kubernetes, focusing on making it easy to learn and develop for Kubernetes."
I would not advise exposing services running within minikube outside of the developer machine running minikube. You may be able to find some stackoverflow with workaround for this, but this is not the intended usage of minikube. 

I think you should try to set up the kubernetes cluster on your server in a different way. 
Since you mentioned your are on a ubuntu server, you could try https://microk8s.io/ since it is by canonical as well, and then install Theia Cloud there.

**Reply by inlann** [reply]
Really appreciate! Does the Getting Start demo use Terraform to invoke Helm chart to deploy the Theia Cloud in minikube? Sorry, I'm new to the Terraform. Should I use Helm chart to deploy the Theia Cloud in a kubernetes cluster directly?

**Reply by jfaltermeier** [reply]
So the GKE and Minikube Getting Started terraform configurations (https://github.com/eclipsesource/theia-cloud/blob/main/terraform/configurations/minikube_getting_started/minikube_getting_started.tf https://github.com/eclipsesource/theia-cloud/blob/main/terraform/configurations/gke_getting_started/gke_getting_started.tf) both use a cluster module (just for the kubernetes cluster creation), a helm module for installation all dependencies and Theia Cloud using helm (https://github.com/eclipsesource/theia-cloud/blob/f984b4b406b57b9f68c6c26f6e52d3f53d0965dd/terraform/configurations/minikube_getting_started/minikube_getting_started.tf#L83) and a keycloak module (for setting up the Realm within the keycloak instance installed in the helm step)

The helm module itself is here: https://github.com/eclipsesource/theia-cloud/blob/main/terraform/modules/helm/main.tf
The `"helm_release"` resources are the helm installations. You should be able to get the repository, version, chart-names, etc. from there if you want to install via helm directly. 

Terraform has the benefit that when you change something in your configurations later on (or want to recreate the whole infrastructure), it is easier to just deploy those changes, because terraform does it for you. When you've manually installed the helm charts it is more error-prone to change things later on.
